Adding physical switches list and summary pages

saulotoledo opened the pull request:
Adds the physical switches show page.

**Depends on:**

- ~~[https://github.com/ManageIQ/manageiq/pull/17216](https://github.com/ManageIQ/manageiq/pull/17216)~~

![switch_summary](https://user-images.githubusercontent.com/5031156/39888127-52718d7e-546a-11e8-9839-dfa4f4bcccc6.png)
![switches_list](https://user-images.githubusercontent.com/5031156/39888124-522ad5dc-546a-11e8-83fb-95f55f6c6354.png)
![switches_menu](https://user-images.githubusercontent.com/5031156/39888125-524f3e7c-546a-11e8-9202-555a1a3f9c71.png)

saulotoledo commented:
@martinpovolny I merged the previous PR here. Now we also have the physical switch summary page. Here is some test data:

https://gist.github.com/saulotoledo/5264bfc1da1ed88b9206d05501982dbf

@skateman, I had to change [this line](https://github.com/saulotoledo/manageiq-ui-classic/blob/b291543049cf59ebb9f65f22f062ad06bf47f620/app/helpers/quadicon_helper.rb#L97) because the base model is `Switch`, and I had to use `PhysicalSwitch`.

saulotoledo commented:
@miq-bot remove_label unmergeable

skateman commented:
@miq-bot add_reviewer @skateman

skateman commented on app/decorators/physical_switch_decorator.rb:
```diff
@@ -0,0 +1,42 @@
+        :tooltip => 'Ports'
+      },
+      :top_right    => {
+        :tooltip => power_state.try(:downcase),
```
Just being curious, is there really a powerstate for a physical switch? :confused:

saulotoledo replied:
Yes, the LXCA returns information about the switch power state. @rodneyhbrown7, do you have more information about?

saulotoledo replied:
I talked to @rodneyhbrown7 and he said that it is ok to have a powerstate for switches.

skateman commented on app/decorators/physical_switch_decorator.rb:
```diff
@@ -0,0 +1,42 @@
+    when "Warning"  then "100/warning.png"
+    else "svg/healthstate-unknown.svg"
+    end
+  end
```
These are okay for now, just note for myself that I have to take a look at this location when converting these to the fonticon+color+background combo in the future.

skateman commented on app/decorators/physical_switch_decorator.rb:
```diff
@@ -0,0 +1,42 @@
+  def single_quad
+    {
+      :fonticon => fonticon
+    }
```
I think you're missing the tooltip here.

saulotoledo replied:
Done. I will update in the next minutes.

skateman commented on app/decorators/physical_switch_decorator.rb:
```diff
@@ -0,0 +1,42 @@
+  def quadicon
+    {
+      :top_left     => {
+        :text    => (hardware.physical_ports ? hardware.physical_ports.size : 0),
```
What about calling `try` instead? It might return a `nil`, but that can be converted to zero using `to_i`:
```ruby
hardware.physical_ports.try(:size).to_i
```

saulotoledo replied:
Done. I will update in the next minutes.

skateman commented:
@miq-bot add_label pending core

skateman commented on app/helpers/quadicon_helper.rb:
```diff
@@ -94,7 +94,7 @@ def self.settings_key(klass, layout)
       end
     else
-      klass.base_model.name.underscore.to_sym
+      klass.name.demodulize.underscore.to_sym
```
Can't be rather the `base_model` definition changed in the `PhysicalSwitch`? Not sure if possible, just asking.

saulotoledo replied:
If we change the base model we have problems with the database. We are using the switches table to store the data. If we define, Rails will try to find the `physical_switches` table.

skateman replied:
I'm just afraid that you're changing this for every model that doesn't fit into the if above, could you please rather do an elsif and keep the else branch if possible?

saulotoledo replied:
Ok. Give me some time to try and I will update that.

saulotoledo replied:
Hi @skateman. What about the following code?

```ruby
elsif klass.base_model.name != klass.name.demodulize
  klass.name.demodulize.underscore.to_sym
else
  klass.base_model.name.underscore.to_sym
end
```

I am not sure if that is really necessary, but I will trust your judgement. If you agree with that proposal I will update the PR with that.

skateman replied:
What if you would store the quadicon setting for the `Switch` model instead? Then you could use the `base_model` and for regular switches it would just fall back to `single_quad` if really necessary.

saulotoledo replied:
Base switches do not have several information from physical switches as power state and health state. I think we can't move the quadicons to them.

skateman replied:
Not the quadicons, just the setting for having a quad or not.

saulotoledo replied:
Hi @skateman, I was playing around with your proposal, but I would like your opinion about something. In order to move the configuration to Switches, I need to change the configuration at `app/views/configuration/_ui_1.html.haml`. It is a good idea to add the role checking, but there is no role defined for Switches yet, only for Physical Switches. I don't think it is a good idea to create roles for Switches right now because we will have the features and the roles in the Core project, and there is no project using them. Also, we will have to check both permissions every time when we want to access Physical Switches (if the user has permissions for Switches and for Physical Switches)

Another solution is to use `true` in the role checking value as we have for `VM`, but the option will always be there. And we will have the problem stated before if someone update that line to have a role to control switches in another part of the project.

The second solution is easy, but I am not sure if it is a good idea if other providers start to use Switches in some place. Adding a role there will break the Physical Switches quadicons.

Thanks in advance. I will wait for your comments about.

skateman replied:
I see no problem in using the role associated to physical switches when it's just about displaying the UI element. The switches have no roles and not even quadicons, so you're completely safe here. Anyway this screen is growing with new datatypes and should be redesigned/refactored a little bit, so you don't even have to feel bad for adding extra complexity. Just please add a comment here that the model is the same, but the role is different and you're good to go :wink:

saulotoledo replied:
As we discussed at Gitter today earlier, I will wait for your return before updating the code. Here are some outputs that can be helpful:

```
irb(main):007:0> Switch.base_model.name
=> "Switch"
irb(main):008:0> PhysicalSwitch.base_model.name
=> "Switch"
irb(main):009:0> Switch.name
=> "Switch"
irb(main):010:0> Switch.name.demodulize
=> "Switch"
irb(main):011:0> PhysicalSwitch.name.demodulize
=> "PhysicalSwitch"
irb(main):012:0> PhysicalServer.name.demodulize
=> "PhysicalServer"
```

skateman replied:
@saulotoledo it's not always the same, demodulize basically removes the `Module::` prefixes and it'ts totally allowed to have a `base_model` that has a `Module::` prefix is totally possible.

saulotoledo replied:
@skateman I've added the ugly `elsif` for now then, and we can handle that in the future as we agreed. Let me know if there are some other changes to do.

saulotoledo commented:
@skateman I added the configuration as you requested (the color for the single quad icon is blue, is that right?). I also added the toolbars as a separated commit on top of the code. Could you take a look there too? Thank you for your help with this commit!

skateman commented:
@saulotoledo yes, single quads should be blue.

martinpovolny commented:
Please, rebase.

saulotoledo commented:
@martinpovolny @skateman Rebase done.

skateman commented:
@saulotoledo I think your deduplication caused a new merge conflict and somehow Travis is failing, can you please rebase it again?

saulotoledo commented:
@skateman New rebase done. :)

skateman commented on app/decorators/physical_switch_decorator.rb:
```diff
@@ -0,0 +1,43 @@
+    {
+      :top_left     => {
+        :text    => hardware.physical_ports.try(:size).to_i,
+        :tooltip => 'Ports'
```
Missing translation?

saulotoledo replied:
fixed

skateman commented on spec/controllers/physical_switch_controller_spec.rb:
```diff
@@ -0,0 +1,59 @@
+  end
+
+  describe "#show_list" do
+    before(:each) do
```
`:each` is redundant here

saulotoledo replied:
fixed

skateman commented on app/decorators/physical_switch_decorator.rb:
```diff
@@ -0,0 +1,43 @@
+      }.merge(QuadiconHelper.machine_state(power_state)),
+      :bottom_left  => {
+        :fonticon => fonticon,
+        :tooltip  => _(type),
```
Umm, this is I think a little bit too much, you need the gettext calls only around the strings. But maybe @mzazrivec or @martinpovolny disagrees, so asking them here...

saulotoledo replied:
@skateman These strings are translated in the list, so I thought that was a good idea to put the translation here too. But I think the translations are not working in the quadicons code. I was about to investigate that, but since it is another issue, I did not investigate yet.

@mzazrivec type is `ManageIQ::Providers::Lenovo::PhysicalInfraManager::PhysicalSwitch`.

If you decide to keep the translation here I will update the other Physical Infra pages in another PR.

mzazrivec replied:
I think instead of `_(type)` you want to use `ui_lookup(:model => type)` here.

saulotoledo replied:
@mzazrivec @skateman Yes, that works. Can I update the code? If yes I will also create another PR for the others physical infra elements. And about the translation at health_state in this same code?

saulotoledo commented:
Inspired by [this comment](https://github.com/ManageIQ/manageiq-ui-classic/pull/3503/files#r186416065) from @martinpovolny in the [racks PR](https://github.com/ManageIQ/manageiq-ui-classic/pull/3503/) I've added some additional tests to this code.

@mzazrivec @skateman I updated the translation for the type. Let me know if you want to keep that translation. In my opinion it is a good idea. If ok, I will open the PRs for the other physical infra elements. Please let me know about the health_state translation too.

skateman commented on app/controllers/physical_switch_controller.rb:
```diff
@@ -0,0 +1,48 @@
+    # pressing the browser's back arrow button to return to the show_list
+    # page while on the Switches' show page. Disabling the cache
+    # causes the page and its session variables to actually be reloaded.
+    disable_client_cache
```
We don't support the browser's native back arrow button, so I don't think this is necessary.

saulotoledo replied:
Done. I will update the repository soon.

martinpovolny commented:
The textual summary part looks good :+1: 

Nice tests! But the test coverage still could be better ;-)

Can you, please, add some checks that the correct toolbars are rendered in the corresponding pages?
Example: https://github.com/ManageIQ/manageiq-ui-classic/pull/3857/files#diff-a32fdcf11645e7b22301a840fd2d6d7eR285

saulotoledo commented:
@martinpovolny @skateman Changes ok.

skateman commented:
Got an error:
```ruby
Error caught: [NoMethodError] undefined method `physical_ports' for nil:NilClass
/home/skateman/Repositories/ManageIQ/manageiq-ui-classic/app/decorators/physical_switch_decorator.rb:9:in `quadicon'
```
